doc/cabal-package.rst

```diff
@@ -1477,8 +1477,22 @@ system-dependent values for these fields.
 
     Version constraints use the operators ``==, >=, >, <, <=`` and a
     version number. Multiple constraints can be combined using ``&&`` or
-    ``||``. If no version constraint is specified, any version is
-    assumed to be acceptable. For example:
+    ``||``.
+    
+    .. Note::
+
+       Even though there is no ``/=`` operator, by combining operators we can
+       skip over one or more versions, to skip a deprecated version or to skip
```

Comment:
Yes, in the [Local no-index repositories](https://cabal.readthedocs.io/en/stable/config.html#local-no-index-repositories) section but that seems a heavy weight way of doing manual deprecations. It doesn't say exactly how to use the `preferred-versions` file to do deprecations but maybe it is the same as I've shown here.

It is quite common to find deprecations on hackage. For instance, 6 out of 9 [ki](https://hackage.haskell.org/package/ki) versions are deprecated and [aeson](https://hackage.haskell.org/package/aeson) has quite a few too.

Reply:
I added a paragraph about hackage `/package-name/preferred` and `/deprecated` links.

Reply:
Even if the solver by itself skips deprecated versions (got a reference for where it does this?), this we can also do ourselves if we want to be super careful (or during debugging to skip a version).

Reply:
Does the solver only view preferred and deprecated versions as "soft constraints"?

https://github.com/haskell/cabal/blob/054108f75c902cc9e2d2b6115bf2fe75707abfb1/cabal-install-solver/src/Distribution/Solver/Types/PackagePreferences.hs#L9-L18

Reply:
> Even if the solver by itself skips deprecated versions (got a reference for where it does this?)

e.g. https://hackage.haskell.org/package/aeson/preferred:

> Deprecating a version adds a range which excludes just that version.

It's worrying that the manual doesn't appear to have this info, though.

> > Are deprecated versions explained anywhere?
>
> in the [Local no-index repositories](https://cabal.readthedocs.io/en/stable/config.html#local-no-index-repositories) section

That's a little surprising, but oh well. Anyways, I'm not convinced that this is a good place to mention deprecated versions, because, I think, it's the whole point of deprecation that the user doesn't have to worry about them and manually add the exclusions. But maybe someone more experienced would be able to chime in...

Reply:
It is good to know how to deprecate stuff as a package author especially for anyone that publishes to hackage. While hackage provides a from for package maintainers with checkboxes for deprecated and freeform entry for preferred versions, cabal should document this too shouldn't it?

I'd like to keep the hackage deprecations note within a note (and I wish I'd known about them sooner).